src/math/spherical_harmonics.cc

```diff
@@ -181,55 +176,215 @@ namespace rascal {
                 assoc_legendre_polynom(angular_l, m_count) * INV_SQRT_TWO;
           } else {
-            // uses symmetry of spherical harmonics,
-            // careful with the storage order
-            harmonics(lm_base + angular_l + m_count) =
+            harmonics(l_block_index + angular_l + m_count) =
                 assoc_legendre_polynom(angular_l, m_count) *
                 cos_sin_m_phi(m_count, 0);
-            harmonics(lm_base + angular_l - m_count) =
+            harmonics(l_block_index + angular_l - m_count) =
                 assoc_legendre_polynom(angular_l, m_count) *
                 cos_sin_m_phi(m_count, 1);
           }  // if (m_count == 0)
         }    // for (m_count in [0, l])
-        lm_base += 2 * angular_l + 1;
+        l_block_index += (2 * angular_l + 1);
       }  // for (l in [0, lmax])
+      // harmonics *= std::sqrt(2.0);
       return harmonics;
     }  // compute_spherical_harmonics()
-  }    // namespace math
+
+    /**
+     * Compute a full set of spherical harmonics and their Cartesian gradients
+     *
+     * Spherical harmonics are defined as described in
+     * math::compute_spherical_harmonics().  Gradients are defined with respect
+     * to motion of the central atom, which is the opposite sign of the usual
+     * definition with respect to the _arguments_ of the Y_l^m.  The actual
+     * Cartesian gradients include an extra factor of 1/r that is not included
+     * here; the rest is independent of radius.
+     *
+     * @param direction Unit vector giving the angles (arguments for the Y_l^m)
+     *
+     * @param max_angular Compute up to this angular momentum number (l_max)
+     *
+     * @return  (Eigen)matrix containing the results.
+     *          Sized 4 by (l_max+1)^2; the first index collects the
+     *          value of the harmonic and the x, y, and z gradient components.
+     *          The second index collects l and m quantum numbers, stored in
+     *          compact format (m varies fastest, from -l to l, and l from 0 to
+     *          l_max).
+     *
+     * @warning This function will access the associated Legendre polynomials
+     *          for m=l+1 and assume they are equal to zero.  The implementation
+     *          of the polynomials in this file respects this convention.
+     *
+     * @todo Add an option to switch off the computation of gradients, so this
+     *       function becomes equivalent to math::compute_spherical_harmonics()
+     */
+    Matrix_t compute_spherical_harmonics_derivatives(
+        const Eigen::Ref<const Eigen::Vector3d> & direction,
+        size_t max_angular) {
+      using std::pow;
+      using std::sqrt;
+      Eigen::Vector3d my_direction;
+      if (std::abs((direction[0] * direction[0] + direction[1] * direction[1] +
+                    direction[2] * direction[2]) -
+                   1.0) > math::dbl_ftol) {
+        std::cerr << "Warning: math::compute_spherical_harmonics_derivatives()";
```

Comment:
Mostly for debugging, to signal that the function is essentially being used the wrong way. But rather than failing, I chose to keep it robust and correct the error with an inefficiency warning. This is the kind of behaviour I'd expect from low-level functionality like this anyway.